CLW/CL/CLW.cl

```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
         }
     }
 }
+
+// --------------------- ATOMIC OPERTIONS ------------------------
+
+#define DEFINE_ATOMIC(operation)\
+    inline void atomic_##operation##_float(volatile __global float* addr, float value)\
+    {\
+        union{\
+        unsigned int u32;\
+        float        f32;\
+        } next, expected, current;\
+        current.f32 = *addr;\
+        do{\
+            expected.f32 = current.f32;\
+            next.f32 = operation(expected.f32, value);\
+            current.u32 = atomic_cmpxchg((volatile __global unsigned int *)addr,\
+                expected.u32, next.u32);\
+        } while (current.u32 != expected.u32);\
+    }
+
+#define DEFINE_ATOMIC_FLOAT3(operation)\
+    inline void atomic_##operation##_float3(volatile __global float3* addr, float3 value)\
+    {\
+        volatile __global float* p = (volatile __global float*)addr;\
+        atomic_##operation##_float(p, value.x);\
+        atomic_##operation##_float(p + 1, value.y);\
+        atomic_##operation##_float(p + 2, value.z);\
+    }
+
+inline void atomic_max_int(volatile __global int* addr, int value)
+{
+    atomic_max(addr, value);
+}
+
+inline void atomic_min_int(volatile __global int* addr, int value)
+{
+    atomic_min(addr, value);
+}
+
+// --------------------- HELPERS ------------------------
+
+#define DEFINE_ASSIGN_OPERATOR(type)\
+    inline void assign_##type(__local type* addr, type value)\
+    {\
+        *addr = value;\
+    }
+
+inline void assign_float3(__local float3* addr, float3 value)
+{
+    (*addr).xyz = value.xyz;
+}
+
+inline int divide_int(int dividend, int divider)
+{
+    return dividend / (divider != 0 ? divider : 1);
+}
+
+inline float divide_float(float dividend, float divider)
```

Comment:
This divide operations look weird, why IEEE floating point division is not enough?

Reply:
Deleted